Implement EXT_disjoint_timer_query

lexaknyazev opened the pull request:
#### dad993a38b597377683cd4cb437018e444819506
<pre>
Implement EXT_disjoint_timer_query
<a href="https://bugs.webkit.org/show_bug.cgi?id=129090">https://bugs.webkit.org/show_bug.cgi?id=129090</a>

Reviewed by Kimmo Kinnunen.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMConvertWebGL.cpp:
(WebCore::convertToJSValue):
* Source/WebCore/html/canvas/EXTDisjointTimerQuery.cpp: Added.
(WebCore::EXTDisjointTimerQuery::EXTDisjointTimerQuery):
(WebCore::EXTDisjointTimerQuery::getName const):
(WebCore::EXTDisjointTimerQuery::supported):
(WebCore::EXTDisjointTimerQuery::createQueryEXT):
(WebCore::EXTDisjointTimerQuery::deleteQueryEXT):
(WebCore::EXTDisjointTimerQuery::isQueryEXT):
(WebCore::EXTDisjointTimerQuery::beginQueryEXT):
(WebCore::EXTDisjointTimerQuery::endQueryEXT):
(WebCore::EXTDisjointTimerQuery::queryCounterEXT):
(WebCore::EXTDisjointTimerQuery::getQueryEXT):
(WebCore::EXTDisjointTimerQuery::getQueryObjectEXT):
* Source/WebCore/html/canvas/EXTDisjointTimerQuery.h: Added.
* Source/WebCore/html/canvas/EXTDisjointTimerQuery.idl: Added.
* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGLAny.h:
* Source/WebCore/html/canvas/WebGLExtension.h:
* Source/WebCore/html/canvas/WebGLRenderingContext.cpp:
(WebCore::WebGLRenderingContext::~WebGLRenderingContext):
(WebCore::WebGLRenderingContext::getExtension):
(WebCore::WebGLRenderingContext::getSupportedExtensions):
(WebCore::WebGLRenderingContext::getInt64Parameter):
(WebCore::WebGLRenderingContext::addMembersToOpaqueRoots):
* Source/WebCore/html/canvas/WebGLRenderingContext.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::getParameter):
(WebCore::WebGLRenderingContextBase::extensionIsEnabled):
(WebCore::WebGLRenderingContextBase::loseExtensions):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/html/canvas/WebGLTimerQueryEXT.cpp: Added.
(WebCore::WebGLTimerQueryEXT::create):
(WebCore::WebGLTimerQueryEXT::~WebGLTimerQueryEXT):
(WebCore::WebGLTimerQueryEXT::WebGLTimerQueryEXT):
(WebCore::WebGLTimerQueryEXT::deleteObjectImpl):
* Source/WebCore/html/canvas/WebGLTimerQueryEXT.h: Added.
* Source/WebCore/html/canvas/WebGLTimerQueryEXT.idl: Added.
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::createQueryEXT):
(WebCore::GraphicsContextGLANGLE::deleteQueryEXT):
(WebCore::GraphicsContextGLANGLE::isQueryEXT):
(WebCore::GraphicsContextGLANGLE::beginQueryEXT):
(WebCore::GraphicsContextGLANGLE::endQueryEXT):
(WebCore::GraphicsContextGLANGLE::queryCounterEXT):
(WebCore::GraphicsContextGLANGLE::getQueryiEXT):
(WebCore::GraphicsContextGLANGLE::getQueryObjectiEXT):
(WebCore::GraphicsContextGLANGLE::getQueryObjectui64EXT):
(WebCore::GraphicsContextGLANGLE::getInteger64EXT):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(createQueryEXT):
(deleteQueryEXT):
(isQueryEXT):
(beginQueryEXT):
(endQueryEXT):
(queryCounterEXT):
(getQueryiEXT):
(getQueryObjectiEXT):
(getQueryObjectui64EXT):
(getInteger64EXT):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createQueryEXT):
(WebKit::RemoteGraphicsContextGLProxy::deleteQueryEXT):
(WebKit::RemoteGraphicsContextGLProxy::isQueryEXT):
(WebKit::RemoteGraphicsContextGLProxy::beginQueryEXT):
(WebKit::RemoteGraphicsContextGLProxy::endQueryEXT):
(WebKit::RemoteGraphicsContextGLProxy::queryCounterEXT):
(WebKit::RemoteGraphicsContextGLProxy::getQueryiEXT):
(WebKit::RemoteGraphicsContextGLProxy::getQueryObjectiEXT):
(WebKit::RemoteGraphicsContextGLProxy::getQueryObjectui64EXT):
(WebKit::RemoteGraphicsContextGLProxy::getInteger64EXT):

Canonical link: <a href="https://commits.webkit.org/260367@main">https://commits.webkit.org/260367@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/63a2827363044d04566d22dc1ba4cb934a40c285

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108114 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17185 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40982 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117239 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116560 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112003 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18540 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8480 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100318 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113882 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97207 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41912 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95890 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28845 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97336 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10066 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30193 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/96700 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8163 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10787 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7094 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16214 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49784 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105720 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7176 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12368 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26175 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -6770,6 +6770,19 @@ WebGLEnabled:
     WebCore:
       default: false
 
+WebGLTimerQueriesEnabled:
```
Thinking about this, could this be lumped with the "WebGL Draft extensions" and retitling that to "Experimental WebGL extensions"?

lexaknyazev replied:
I think these are quite different. Draft extensions are supposed to be promoted so they are behind the flag for a limited time. Timer queries could be seen as "developer" or "sensitive", meaning that they may forever remain behind a flag on some platforms.

kkinnunen-apple replied:
Right, but the concept of "draft extensions" itself is similar to "developer extensions".

So we might have:

- draft extension 1
- draft extension 2
- developer extension 1
- developer extension 2
- developer extension 3

So we don't want settings:
- WebGL enable draft extensions
- WebGL enable developer extension 1
- WebGL enable developer extension 2
- WebGL enable developer extension 3

This is on the count that that is just too messy, proliferation of settings

So we don't probably want settings:
- WebGL enable draft extensions
- WebGL enable developer extensions

This is on the count that that is just too messy, proliferation of settings. I imagine it's the same set of people enabling both and there's little to be gained for enabling one but not the other.

So we should just have:
- WebGL enable developer extensions

lexaknyazev replied:
> So we should just have:
> - WebGL enable developer extensions

This implies that either:
1. timer query extensions will remain a developer-only feature that is disabled by default, or
2. timer query extensions will be always enabled.

If it is (1), the flags should be merged; if it is (2), the extension exposure should be unconditional.

kkinnunen-apple replied:
I think we should start with 1. However, merging the settings can be done in a separate patch, if that is simpler. I can also do that.

lexaknyazev replied:
Yes, let's defer merging the settings because renaming the "draft extensions" flag would also require updating several tests that rely on that name.

kkinnunen-apple commented on Source/WebCore/html/canvas/EXTDisjointTimerQuery.cpp:
```diff
@@ -0,0 +1,264 @@
+        return;
+
+    if (!query->validate(context->contextGroup(), *context)) {
+        context->synthesizeGLError(GraphicsContextGL::INVALID_OPERATION, "delete", "object does not belong to this context");
```
would the function name be `"deleteQueryEXT"` in the error message?

lexaknyazev replied:
This is a long standing issue that I plan to address separately.

Almost all `delete*` functions call `WebGLRenderingContextBase::deleteObject` that is not aware of the original function name and therefore prints
> delete: object does not belong to this context

To ensure that similar errors produce similar error messages, a few special validation cases follow the same suboptimal pattern for now. A separate PR will fix that for all objects.

kkinnunen-apple replied:
sounds good

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -2959,6 +2959,96 @@ void GraphicsContextGLANGLE::drawBuffersEXT(GCGLSpan<const GCGLenum> bufs)
+    if (!makeContextCurrent())
+        return 0;
+
+    GLuint name;
```
Could we get the out variables all initialized?

lexaknyazev replied:
Done.

kkinnunen-apple reviewed (approved):
Nice work!
very minor notes that could be fixed

lexaknyazev committed:
Implement EXT_disjoint_timer_query
https://bugs.webkit.org/show_bug.cgi?id=129090

Reviewed by Kimmo Kinnunen.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMConvertWebGL.cpp:
(WebCore::convertToJSValue):
* Source/WebCore/html/canvas/EXTDisjointTimerQuery.cpp: Added.
(WebCore::EXTDisjointTimerQuery::EXTDisjointTimerQuery):
(WebCore::EXTDisjointTimerQuery::getName const):
(WebCore::EXTDisjointTimerQuery::supported):
(WebCore::EXTDisjointTimerQuery::createQueryEXT):
(WebCore::EXTDisjointTimerQuery::deleteQueryEXT):
(WebCore::EXTDisjointTimerQuery::isQueryEXT):
(WebCore::EXTDisjointTimerQuery::beginQueryEXT):
(WebCore::EXTDisjointTimerQuery::endQueryEXT):
(WebCore::EXTDisjointTimerQuery::queryCounterEXT):
(WebCore::EXTDisjointTimerQuery::getQueryEXT):
(WebCore::EXTDisjointTimerQuery::getQueryObjectEXT):
* Source/WebCore/html/canvas/EXTDisjointTimerQuery.h: Added.
* Source/WebCore/html/canvas/EXTDisjointTimerQuery.idl: Added.
* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGLAny.h:
* Source/WebCore/html/canvas/WebGLExtension.h:
* Source/WebCore/html/canvas/WebGLRenderingContext.cpp:
(WebCore::WebGLRenderingContext::~WebGLRenderingContext):
(WebCore::WebGLRenderingContext::getExtension):
(WebCore::WebGLRenderingContext::getSupportedExtensions):
(WebCore::WebGLRenderingContext::getInt64Parameter):
(WebCore::WebGLRenderingContext::addMembersToOpaqueRoots):
* Source/WebCore/html/canvas/WebGLRenderingContext.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::getParameter):
(WebCore::WebGLRenderingContextBase::extensionIsEnabled):
(WebCore::WebGLRenderingContextBase::loseExtensions):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/html/canvas/WebGLTimerQueryEXT.cpp: Added.
(WebCore::WebGLTimerQueryEXT::create):
(WebCore::WebGLTimerQueryEXT::~WebGLTimerQueryEXT):
(WebCore::WebGLTimerQueryEXT::WebGLTimerQueryEXT):
(WebCore::WebGLTimerQueryEXT::deleteObjectImpl):
* Source/WebCore/html/canvas/WebGLTimerQueryEXT.h: Added.
* Source/WebCore/html/canvas/WebGLTimerQueryEXT.idl: Added.
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::createQueryEXT):
(WebCore::GraphicsContextGLANGLE::deleteQueryEXT):
(WebCore::GraphicsContextGLANGLE::isQueryEXT):
(WebCore::GraphicsContextGLANGLE::beginQueryEXT):
(WebCore::GraphicsContextGLANGLE::endQueryEXT):
(WebCore::GraphicsContextGLANGLE::queryCounterEXT):
(WebCore::GraphicsContextGLANGLE::getQueryiEXT):
(WebCore::GraphicsContextGLANGLE::getQueryObjectiEXT):
(WebCore::GraphicsContextGLANGLE::getQueryObjectui64EXT):
(WebCore::GraphicsContextGLANGLE::getInteger64EXT):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(createQueryEXT):
(deleteQueryEXT):
(isQueryEXT):
(beginQueryEXT):
(endQueryEXT):
(queryCounterEXT):
(getQueryiEXT):
(getQueryObjectiEXT):
(getQueryObjectui64EXT):
(getInteger64EXT):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createQueryEXT):
(WebKit::RemoteGraphicsContextGLProxy::deleteQueryEXT):
(WebKit::RemoteGraphicsContextGLProxy::isQueryEXT):
(WebKit::RemoteGraphicsContextGLProxy::beginQueryEXT):
(WebKit::RemoteGraphicsContextGLProxy::endQueryEXT):
(WebKit::RemoteGraphicsContextGLProxy::queryCounterEXT):
(WebKit::RemoteGraphicsContextGLProxy::getQueryiEXT):
(WebKit::RemoteGraphicsContextGLProxy::getQueryObjectiEXT):
(WebKit::RemoteGraphicsContextGLProxy::getQueryObjectui64EXT):
(WebKit::RemoteGraphicsContextGLProxy::getInteger64EXT):

Canonical link: https://commits.webkit.org/260367@main
